Interpolater.C

```diff
@@ -0,0 +1,181 @@
+#include "mpi.h"
+
+/* Use C++11 built-in shared pointers if available; else fallback to Boost. */
+#if __cplusplus >= 201103L
+#include <memory>
+#else
+#include <boost/shared_ptr.hpp>
+#endif
+
+using namespace std;
+
+namespace CAROM {
+
+Interpolater::Interpolater(std::vector<Vector*> parameter_points,
+                           std::vector<Matrix*> rotation_matrices,
+                           int ref_point,
+                           double epsilon)
+{
+    CAROM_VERIFY(parameter_points.size() == rotation_matrices.size());
+    CAROM_VERIFY(parameter_points.size() > 0);
+
+    // Get the rank of this process, and the number of processors.
+    int mpi_init;
+    MPI_Initialized(&mpi_init);
+    if (mpi_init == 0) {
+        MPI_Init(nullptr, nullptr);
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &d_rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &d_num_procs);
+
+    d_parameter_points = parameter_points;
+    d_rotation_matrices = rotation_matrices;
+    d_ref_point = ref_point;
+    d_lambda_T = NULL;
+    d_epsilon = epsilon;
+}
+
+std::vector<double> Interpolater::obtainRBF(Vector* point)
+{
+    // Inverse quadratic RBF
+    std::vector<double> inv_q;
+    for (int i = 0; i < d_parameter_points.size(); i++)
+    {
+        Vector diff;
+        point->minus(*d_parameter_points[i], diff);
+        double res = 1.0 / (1.0 + std::pow(d_epsilon*diff.norm(), 2));
+        inv_q.push_back(res);
+    }
+
+    return inv_q;
+}
+
+std::vector<Matrix*> obtainRotationMatrices(std::vector<Vector*> parameter_points,
+        std::vector<Matrix*> bases,
+        int ref_point)
+{
+    // Get the rank of this process, and the number of processors.
+    int mpi_init;
+    int rank;
+    int num_procs;
+    MPI_Initialized(&mpi_init);
+    if (mpi_init == 0) {
+        MPI_Init(nullptr, nullptr);
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
+
+    std::vector<Matrix*> rotation_matrices;
+
+    // Obtain the rotation matrices to rotate the bases into
+    // the same generalized coordinate space.
+    for (int i = 0; i < parameter_points.size(); i++)
+    {
+        CAROM_VERIFY(bases[i]->numRows() == bases[ref_point]->numRows());
+        CAROM_VERIFY(bases[i]->numColumns() == bases[ref_point]->numColumns());
+        CAROM_VERIFY(bases[i]->distributed() == bases[ref_point]->distributed());
+
+        // If at ref point, the rotation_matrix is the identity matrix
+        // since the ref point doesn't need to be rotated.
+        if (i == ref_point)
+        {
+            Matrix* identity_matrix = new Matrix(bases[i]->numColumns(), bases[i]->numColumns(), false);
+            for (int j = 0; j < identity_matrix->numColumns(); j++) {
+                identity_matrix->item(j, j) = 1.0;
+            }
+            rotation_matrices.push_back(identity_matrix);
+            continue;
+        }
+
+        Matrix* basis_mult_basis = bases[i]->transposeMult(bases[ref_point]);
```

Comment:
Isn't this always a non-distributed, small matrix? If so, is the SVD computation really done in parallel?

Reply:
Yes, transposeMult always returns a undistributed matrix, so the SVD computation is not done in parallel.

Reply:
The basis matrices are large and distributed row-wise, so the transposeMult is performed in parallel and produces a small matrix of size equal to the number of basis vectors. This is fine, but my question was about whether the SVD should be done in parallel by scalapack. If it is small, then lapack can do the SVD in serial. In libROM we use scalapack for SVD of a large distributed matrix (e.g. `StaticSVD`), not for small matrices like this.

Reply:
Is serial SVD with scalapack is particularly slower than lapack?

Reply:
Not sure about performance, but at least the code for lapack seems simpler and shorter (compare QDEIM.C and Interpolator.C). Maybe the SVD can be put into a function to be more compact and avoid repetition.